oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,402 @@
+* A version constraint needs to be added to prevent upgrading to a
+  backwards-incompatible release of a required package until appropriate code
+  changes can be made.
+* The code has been updated to support a newer dependency package version
+  which was previously blocked by a version constraint.
+
+Whenever a developer needs to make a deliberate change to the repository's
+Python package dependencies, they should do the following:
+
+1. Make the appropriate changes to the ``*.in`` files.
+2. Run ``make upgrade`` to regenerate the detailed requirements files.
+3. For each package for which the pinned version is changing in the ``*.txt``
+   requirements files, look at its changelog to make sure that there
+   are no problematic backwards-incompatible changes.  If there are, add
+   a version constraint to one of the ``.in`` files to prevent it from being
+   upgraded to that release, run ``make upgrade`` again, and file a ticket
+   briefly describing the change that needs to be made in order to upgrade
+   that package further.  Similarly, if there are new features that the code
+   depending on that package should start taking advantage of, file tickets
+   explaining what should be done.
+4. Check in all of the changed requirements files and wait for the automated
+   test results.  If one of the upgrades caused unexpected problems, follow
+   the same process as if a backwards-incompatible change had been spotted in
+   the changelog (add a version constraint, ``make upgrade``, file a ticket).
+
+Manually editing the ``make upgrade`` output files or only running
+``pip-compile`` on a single file should generally be avoided, since it risks
+failing to account for changes in indirect dependencies or making the
+different requirement files fall out of sync with each other.  And in general,
+we would rather err on the side of using newer versions of dependencies than
+strictly necessary, rather than avoiding upgrades for fear of breaking things.
+If the developer is not confident of their ability to assess whether a change
+to the dependencies is appropriate, they should seek assistance from other
+developers who are either more experienced or more familiar with that
+repository.
+
+Installing Dependencies from URLs
+---------------------------------
+
+As noted above, you should generally avoid installing requirements from a URL
+or local directory instead of PyPI.  But there are a few circumstances where
+it can be appropriate:
+
+* You need to test a release candidate of the dependency to make sure it will
+  work with your code.
+* You critically need a fix for a package which has not yet been included in
+  a release, and you cannot arrange for a release to be made in a timely
+  manner.
+
+In most other circumstances, the package should be added to PyPI instead.  If
+you do need to include a package at a URL, it should be editable (start with
+"-e ") and have both the package name and version specified (end with
+"#egg=NAME==VERSION").
```

Comment:
This seems like a worthwhile place to have an example of a fully-baked url include line.

Reply:
Sure, I'll add an example.